[ClangImporter] Forward generic parameters while importing generic @c…

xedin opened the pull request:
…ompatibility_alias

When importing @compatibility_alias declarations check if underlying declaration
is generic and if so, forward generic environment and generic parameters (if any)
to newly created typealias declaration, otherwise there is going to be a mismatch
between type associated with typealias and its declaration which leads to crashes.

Resolves: rdar://problem/39849926

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ClangImporter] Forward generic parameters while importing generic @compatibility_alias

When importing @compatibility_alias declarations check if underlying declaration
is generic and if so, forward generic environment and generic parameters (if any)
to newly created typealias declaration, otherwise there is going to be a mismatch
between type associated with typealias and its declaration which leads to crashes.

Resolves: rdar://problem/39849926

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4966,6 +4966,12 @@ namespace {
+      if (auto *GTD = dyn_cast<GenericTypeDecl>(typeDecl)) {
+        typealias->setGenericEnvironment(GTD->getGenericEnvironment());
+        if (GTD->isGeneric())
+          typealias->setGenericParams(GTD->getGenericParams()->clone(dc));
```
Doesn't the typealias have to be the DeclContext?

xedin replied:
Do you mean `GenericContext`? I'm checking for `GenericTypeDecl` here because `typeDecl` already is a `TypeDecl` so not sure if it makes sense to check lower than that?

jrose-apple replied:
No, sorry, I mean in the call to `clone`.

xedin replied:
Ah, sure, I'll change that!

jrose-apple commented on test/ClangImporter/Inputs/custom-modules/ObjCIRExtras.h:
```diff
@@ -60,8 +60,12 @@
 @interface SwiftNameTestErrorSub : SwiftNameTestError
 @end
 
+@interface SwiftGenericNameTest<T> : NSObject
```
Please include a test where T is constrained (either to a class or to a protocol or both).

xedin replied:
I think particular case T is constrained to NSObject, I have a test which tried to pass `Int` to it and fails, but I'll add more too.

xedin replied:
I've added one more class which has T constrained to NSNumber *

jrose-apple replied:
I believe it's just constrained to AnyObject. NSNumber is good enough!

jrose-apple reviewed (commented):
Thanks, but this isn't the only place where the importer creates typealiases for generic types. The other is renames; see `SwiftDeclConverter::importCompatibilityTypeAlias` (unrelated to `@compatibility_alias`).

xedin committed:
[ClangImporter] Forward generic parameters to renamed typealias

While trying to import declaration which requires renaming forward
generic parameters (if any) to newly created typealias.

xedin commented:
@jrose-apple I've made changes to `importCompatibilityTypeAlias` and fixed `clone` to use `GTD` instead of `dc`, WDYT?

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4966,6 +4966,12 @@ namespace {
+      if (auto *GTD = dyn_cast<GenericTypeDecl>(typeDecl)) {
+        typealias->setGenericEnvironment(GTD->getGenericEnvironment());
+        if (GTD->isGeneric())
+          typealias->setGenericParams(GTD->getGenericParams()->clone(GTD));
```
This is still wrong; it's supposed to be the *new* DeclContext, not the old one. That's `typealias`.

xedin replied:
I get it now, sorry - it has to use `typealias` because that's the scope for the new generic parameters, sorry about this...

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -5204,7 +5198,15 @@ Decl *SwiftDeclConverter::importCompatibilityTypeAlias(
+  if (GTD && !isa<ProtocolDecl>(GTD)) {
+    alias->setGenericEnvironment(GTD->getGenericEnvironment());
+    if (GTD->isGeneric())
+      alias->setGenericParams(GTD->getGenericParams()->clone(GTD));
```
(here too)

jrose-apple reviewed (commented):
Tests look good, code still a bit off!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test source compatibility
